tests/unit/python/foglamp/plugins/south/coap_listen/test_coap_listen.py

```diff
@@ -0,0 +1,224 @@
+                 'description': 'Python module name of the plugin to load',
+                 'type': 'string',
+                 'default': 'coap_listen'
+            },
+            'port': {
+                'description': 'Port to listen on',
+                'type': 'integer',
+                'default': str(unused_port),
+            },
+            'uri': {
+                'description': 'URI to accept data on',
+                'type': 'string',
+                'default': 'sensor-values',
+            }
+    }
+
+_NEW_CONFIG = {
+    'plugin': {
+         'description': 'Python module name of the plugin to load',
+         'type': 'string',
+         'default': 'coap_listen'
+    },
+    'port': {
+        'description': 'Port to listen on',
+        'type': 'integer',
+        'default': '1234',
+    },
+    'uri': {
+        'description': 'URI to accept data on',
+        'type': 'string',
+        'default': 'sensor-values',
+    }
+}
+
+
+def test_plugin_info():
+    assert coap_listen.plugin_info() == {
+            'name': 'CoAP Server',
+            'version': '1.0',
+            'mode': 'async',
+            'type': 'south',
+            'interface': '1.0',
+            'config': _DEFAULT_CONFIG(5683)
+    }
+
+def test_plugin_init(unused_port):
+    config = _DEFAULT_CONFIG(unused_port())
+    assert coap_listen.plugin_init(config) == config
+
+def test_plugin_start(mocker, unused_port):
+    # GIVEN
+    config = copy.deepcopy(_DEFAULT_CONFIG(unused_port()))
+    config['port']['value'] = config['port']['default']
+    config['uri']['value'] = config['uri']['default']
+
+    log_exception = mocker.patch.object(coap_listen._LOGGER, "exception")
+    log_info = mocker.patch.object(coap_listen._LOGGER, "info")
+    assert coap_listen.aiocoap_ctx is None
+
+    # WHEN
+    loop = asyncio.get_event_loop()
+    loop.call_soon(coap_listen.plugin_start(config))
+
+    # THEN
+    assert coap_listen.aiocoap_ctx is not None
+    assert 1 == log_info.call_count
+    calls = [call('CoAP listener started on port {} with uri {}'.format(config['port']['value'], config['uri']['value']))]
+    log_info.assert_has_calls(calls, any_order=True)
+
+def test_plugin_reconfigure(mocker, unused_port):
+    # GIVEN
+    config = copy.deepcopy(_DEFAULT_CONFIG(unused_port()))
+    config['port']['value'] = config['port']['default']
+    config['uri']['value'] = config['uri']['default']
+    mocker.patch.object(coap_listen, '_plugin_stop', return_value=True)
+    log_exception = mocker.patch.object(coap_listen._LOGGER, "exception")
+    log_info = mocker.patch.object(coap_listen._LOGGER, "info")
+
+    # WHEN
+    new_config = coap_listen.plugin_reconfigure(config, _NEW_CONFIG)
+
+    # THEN
+    assert _NEW_CONFIG == new_config
+    assert 2 == log_info.call_count
+    # TODO: log_info.assert_called_with('Restarting HTTP_SOUTH plugin due to change in configuration keys [uri, port, host]')
+
+
+def test_plugin__stop(mocker, unused_port):
```

Comment:
Why not patch aiocoap_ctx.shutdown() and assert that it is called?

Reply:
The point is - we should keep mocking at bare minimum. This will help us in finding fault in code at the earliest and conveniently.